File: C/beadsort/BlackBox/BlackBox.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void beadSort(int* arr, int n) {
    // Find the maximum value in the array
    int max = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }

    // Create a "beads" array
    int** beads = (int**)malloc((max + 1) * sizeof(int*));
    for (int i = 0; i <= max; i++) {
        beads[i] = (int*)calloc(n, sizeof(int));
    }

    // Place beads in the beads array
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < arr[i]; j++) {
            beads[j][i] = 1; // Place a bead
        }
    }

    // Sort the beads
    for (int i = max; i >= 0; i--) {
        int count = 0;
        for (int j = 0; j < n; j++) {
            if (beads[i][j] == 1) {
                count++;
            }
        }
        for (int j = 0; j < count; j++) {
            arr[n - 1 - j] = i; // Set sorted values back to the array
        }
    }

    // Free the allocated memory
    for (int i = 0; i <= max; i++) {
        free(beads[i]);
    }
    free(beads);
}

int main() {
    int n = 1000;
    int* arr = (int*)malloc(n * sizeof(int));

    // Seed the random number generator
    srand(time(NULL));

    // Generate random values between 0 and 100
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % 101; // Random values from 0 to 100
    }

    // Sort the array using bead sort
    beadSort(arr, n);

    // Free the allocated memory
    free(arr);
    
    return 0;
}

File: Python/beadsort/BlackBox/BlackBox.py
import random

def bead_sort(arr):
    max_value = max(arr)
    # Create a list of lists (beads)
    beads = [[0] * len(arr) for _ in range(max_value)]

    # Place beads
    for i in range(len(arr)):
        for j in range(arr[i]):
            beads[j][i] = 1

    # Sum the beads column-wise
    for j in range(len(arr)):
        count = sum(beads[i][j] for i in range(max_value))
        for i in range(count):
            beads[max_value - 1 - i][j] = 1
        for i in range(count, max_value):
            beads[i][j] = 0

    # Read the sorted values back into the array
    for i in range(len(arr)):
        arr[i] = sum(beads[j][i] for j in range(max_value))

def main():
    # Generate 1000 random integers between 0 and 100
    arr = [random.randint(0, 100) for _ in range(1000)]
    
    bead_sort(arr)

if __name__ == "__main__":
    main()